odl/discr/discr_utils.py

```diff
@@ -1339,12 +1339,12 @@ def dual_use_func(x, out=None, **kwargs):
             elif tensor_valued:
                 # The out object can be any array-like of objects with shapes
                 # that should all be broadcastable to scalar_out_shape.
-                results = np.array(out)
-                if results.dtype == object or scalar_in:
+
+                if any(res.shape != scalar_out_shape for res in out) or scalar_in:
                     # Some results don't have correct shape, need to
                     # broadcast
                     bcast_res = []
-                    for res in results.ravel():
+                    for res in out:
```

Comment:
This loops only through the outer dimension. I think we should either remove the broadcasting all together or make it general.